Add missing tests for image block

artemiomorales opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
A follow up to #51278, this adds a test to ensure that the experimental lightbox works as expected when inserting an image from URL.

It also adds other missing checks:
- Inserting an image via URL works and renders the image appropriately on the frontend
- Adding a link to an image will work as expected
- When a lightbox is configured, an image src should only load when the lightbox is opened
- Adding a link to an image will override the lightbox

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
This will help us ensure the image default behavior, as well as lightbox, perform as expected.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
N/A

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
Run the tests locally using `npm run test:e2e:playwright -- image.spec.js`

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
N/A

artemiomorales committed:
Add missing tests for image block

artemiomorales committed:
Add test for image link

artemiomorales commented:
@kevin940726 Again, I was unable to get `page.getByRole()` working in some cases. Would appreciate any suggestions on this front!

artemiomorales commented on test/e2e/specs/editor/blocks/image.spec.js:
```diff
@@ -855,6 +940,89 @@ test.describe( 'Image - interactivity', () => {
+		page,
+	} ) => {
+		await admin.createNewPost();
+		await editor.insertBlock( { name: 'core/image' } );
```
@kevin940726 This test case doesn't make full use of the `beforeAll().` Should it be separated from this test suite?

kevin940726 replied:
What do you mean? 🤔

artemiomorales replied:
Ok, so what I mean to say is, not all test cases under the `Image - Interactivity` test suite require the logic in the setup and teardown methods.

We have two options that I see:
1. Break down the test suite into more discrete groupings so code doesn't run unnecessarily and make it perhaps easier to reason with the code (though having more groupings might actually increase cognitive overload)

2. Leave the tests where they are and trust that devs will be able to deduce that some tests don't require the use of the setup and teardown methods; perhaps leave a comment

Now that I write this out, I'm leaning towards the first option. What do you think?

artemiomorales replied:
I went ahead and reorganized the tests. It should be clearer to anyone who takes a look at these in the future.

artemiomorales commented on test/e2e/specs/editor/blocks/image.spec.js:
```diff
@@ -739,10 +742,90 @@ test.describe( 'Image', () => {
 
 test.describe( 'Image - interactivity', () => {
 	let filename = null;
+	let imageBlock = null;
```
@kevin940726 This is only used in one test case below. Should I define it here at all?

kevin940726 replied:
I'm all for removing it to keep it simple 👍

artemiomorales replied:
Got it, thanks, will revise 👍

artemiomorales commented:
I could use a second opinion on the best way to organize these tests. I added a couple of comments related to that; essentially, the code we use to prepare the environment don't apply in all of the test scenarios.

artemiomorales reviewed (commented):
I could use a second opinion on the best way to organize these tests. I added a couple of comments related to that; essentially, the code we use to prepare the environment don't apply in all of the test scenarios.

kevin940726 commented on test/e2e/specs/editor/blocks/image.spec.js:
```diff
@@ -739,10 +742,90 @@ test.describe( 'Image', () => {
+			.click();
+
+		const form = page.locator(
+			'.block-editor-media-placeholder__url-input-form'
```
Seems like there should be a `dialog` here but it doesn't exist 😅 . I think it's okay to use class names as long as we add a comment above it 👍

artemiomorales replied:
Ok got it! Will add a comment 👍

artemiomorales replied:
Comment added.

artemiomorales committed:
Clean up tests

-Reorganize tests to avoid unnecessary code execution
-Add comments
-Remove extraneous code

artemiomorales commented:
@kevin940726 Changes pushed. Let me know how it looks and if there's anything else I should address!

kevin940726 commented on test/e2e/specs/editor/blocks/image.spec.js:
```diff
@@ -869,63 +929,182 @@ test.describe( 'Image - interactivity', () => {
+			} );
+			await behaviorSelect.selectOption( 'lightbox' );
+
+			//await imageBlock.click();
```
Possibly an oversight?

artemiomorales replied:
Yes, it was, thanks! Fixed.

kevin940726 commented on test/e2e/specs/editor/blocks/image.spec.js:
```diff
@@ -869,63 +929,182 @@ test.describe( 'Image - interactivity', () => {
+			const postId = await editor.publishPost();
+			await page.goto( `/?p=${ postId }` );
+
+			expect(
+				await page.locator( '.wp-lightbox-overlay' ).count()
+			).toEqual( 0 );
```
Possibly using `toBeVisible()` will be more readable here?

artemiomorales replied:
Hmm so `toBeVisible()` would not work here, because the overlay shouldn't appear in the DOM at all.
I added a comment to clarify. I'm not sure that any other syntax would work.

kevin940726 replied:
I see! Thanks for the clarification!

Hmm, I guess now the problem is why do we want to make this assertion? It seems less ideal to me to assert something in a way that end users can't interact with. If we want to test if clicking on the overlay will close the lightbox then we should just write that test instead. If what we really want to test is the visibility of the scrim element then maybe we should select that instead?

These are implementation details and might change, for example, when we decide to use the native `dialog` element instead. Then we would have to refactor the test even though users' interactions haven't changed.

Now that I think about it, why don't we use the native [`::backdrop`](https://developer.mozilla.org/en-US/docs/Web/CSS/::backdrop) pseudo element in the first place? It seems to be supported in all the browsers we support.

If none of the above work, then I guess we can still use [`toBeInViewport()`](https://playwright.dev/docs/api/class-locatorassertions#locator-assertions-to-be-in-viewport)?

---

With all that said, I'm just stating my observations, and I don't think any of it should block this wonderful PR! I'm just curious about the decisions we made around this area because I'm less informed about it 😅.

artemiomorales replied:
> Hmm, I guess now the problem is why do we want to make this assertion?

Thanks for the thoughts!

To clarify, this test is meant to verify the following: When a lightbox is configured for an image, the lightbox markup should NOT appear in the frontend if a link is enabled for the image. The reason we want to test this is that we cannot have both the lightbox and a link work simultaneously — we need to pick one. And so, enabling a link should override the lightbox functionality.

The ideas you’ve suggested may work for other scenarios, but not for this case unfortunately. Would be happy to hear of any other potential approaches.

> Now that I think about it, why don't we use the native ::backdrop pseudo element in the first place? It seems to be supported in all the browsers we support.

I tried using ::backdrop, but for this implementation, we need to set the scrim to be the site background color, so using ::backdrop would have involved the use of CSS variables, and that seemed less straightforward and harder to follow than just setting the background color on a scrim div.

Just let me know if I can clarify anything further 🙏

kevin940726 replied:
Yep! I should really re-read the test title 😅 .

In this case, I think checking user interactions could still work? A simple solution would be to test if the enlarge button still exists.

```js
await expect( page.getByRole( 'button', { name: 'Enlarge image' } ) ).not.toBeInViewport();
// `toBeAttached()` will be better but we don't support that API now. 😞 
```

What I'm trying to do is always trying to target user-visible elements with accessible attributes. In this case, `.wp-lightbox-overlay` is not visible and an implementation detail (unless we document it somewhere though!), but the enlarge button is accessible to screen readers. We can make sure that every time we change the test is because of some user flow changes. LMK if that helps! 🙇 

> so using ::backdrop would have involved the use of CSS variables, and that seemed less straightforward and harder to follow than just setting the background color on a scrim div.

Hmm, I personally don't think it's a problem TBH 😅 . If the spec is suggesting us to do this then I don't see why we should avoid using CSS variables in this case. I think it's just a matter of choosing the variable name? Something like... (not tested)

```html
<!-- in PHP... -->
<div class="wp-lightbox-overlay" style="--wp-lightbox-scrim-color: $background_color;"></div>

<!-- and in SCSS... -->
<style>
.wp-lightbox-overlay::backdrop {
  background-color: var(--wp-lightbox-scrim-color);
}
</style>
```

artemiomorales replied:
>```
>await expect( page.getByRole( 'button', { name: 'Enlarge image' } ) ).not.toBeInViewport();
>// `toBeAttached()` will be better but we don't support that API now. 😞 
>```

@kevin940726 Thanks for this! Agreed that this is more readable — I made this change.

>Hmm, I personally don't think it's a problem TBH 😅 . If the spec is suggesting us to do this then I don't see why we should avoid using CSS variables in this case. I think it's just a matter of choosing the variable name? Something like... (not tested)

Thanks for taking a look at this; we might be able to implement that — however, another issue with using `::backdrop` is that I'm unable to animate it, at least in my testing.

Here is the lightbox using the `scrim` div. Note the background fading in smoothly:

https://github.com/WordPress/gutenberg/assets/5360536/6eaf2f55-eaf2-437d-9f78-bf8b57cf906c

Contrast this with the following `::backdrop` test:

https://github.com/WordPress/gutenberg/assets/5360536/41551b61-3cf6-4bcd-a219-62df3ba7a3b8

As far as I can tell then, using `::backdrop` is not an option.

With this in mind, we could potentially switch the `.wp-lightbox-overlay` div to be a `dialog` element for semantic clarity in the tests and the markup. Note that when I was first implementing the lightbox, I just followed the convention that already existed in the codebase to iterate quickly and get feedback as quickly as possible. 

How the lightbox currently works is modeled on the responsive navigation menu implementation, which does not use the `dialog` _element_ but toggles the `dialog` _role_ on a div based on whether the modal is open. I suppose now is the time for us to revisit whether that implementation was the best approach or not.

So as far as semantic improvements go, we have a few options:

1. Leave the implementation as is
2. Leave the the `dialog` role on the `.wp-lightbox-overlay` whether the div is open or not
3. Switch the `.wp-lightbox-overlay` div to be a `dialog` element

That being said, I'm not sure which of these three approaches is the best option.
- In case of Option 1, I'm not sure why the navigation modal was implemented the way it was to begin with; perhaps there's a reason we're not aware of.
- With Option 2, I feel like this approach makes sense because, even though it would have the `dialog` role at all times, it has `aria-hidden` and `aria-modal="false"` to prevent it from being erroneously read by screenreaders.
- With Option 3, that makes sense as per the book, but we're not using the `dialog` element's other features, like `show()`, `showModal()`, or `::backdrop`, so perhaps it makes more sense to use a div to signal that it's a completely custom implementation.

These are my thoughts. What do you think?

I'm about to start working on more tests for the [lightbox zoom animation](https://github.com/WordPress/gutenberg/pull/51357) in a separate PR, so will be able to make any revisions then.

kevin940726 reviewed (commented):
Thank you! Left some suggestions that will be nice to fix before merging. This is looking awesome though! 🙇

artemiomorales committed:
Remove commented code

artemiomorales committed:
Add clarifying comment

kevin940726 reviewed (approved):
Thank you! ❤️

artemiomorales committed:
Improve syntax for test

Mamaduka commented:
Hey, @artemiomorales 👋 

It looks like the image "lightbox" e2e tests became flaky recently. If you query the label, the first few reports are all related to this feature - https://github.com/WordPress/gutenberg/issues?q=is%3Aissue+is%3Aopen+label%3A%22%5BType%5D+Flaky+Test%22.

artemiomorales commented:
@Mamaduka Thanks for the heads up! Will take a look
